Index and expose in API symbol metadata

kzh opened the pull request:
Edit: This PR is now just about indexing and returning the metadata. We landed supporting regex + symbol straight to master.

This is the other route to implement regex support for symbol search. It avoids storing additional information to support regex search. It's as fast as text search. The caveats are there are more potential candidate matches that are negative. This might not be a hindrance to performance and a reasonable tradeoff to using more memory.

For substring symbol search, we use what Zoekt has originally. It does a regular substring text search and then checks if those matches are within the bounds of a symbol document section. 

For regex symbol search, we run the regex against all symbols in candidate files.

~Note: There is some weird behaviour with `sourcegraph/zoekt` that appears as soon as you remove `o.CTags = ""` in `build/builder.go` before even changing anything else. The results are truncated (or missing some). This might be due to how we limit results per shard?  For this reason, I built this PR on top of `google/zoekt`.~

```
# installing:
$ go install ./cmd/zoekt/
$ go install ./cmd/zoekt-index/

# universal-ctags (w/ interactive mode) is required prior to indexing.

# indexing:
$ zoekt-index ./

# example queries:
$ zoekt 'sym:^IndexBuilder$'
$ zoekt 'sym:^test_.*'
$ zoekt 'sym:.* file:indexbuilder.go'
```

Todo:
- ~store corresponding ctags metadata~
- ~expose the metadata in the zoekt api~

kzh commented on matchtree.go:
```diff
@@ -141,6 +141,95 @@ type branchQueryMatchTree struct {
+		fileStart = t.fileEndRunes[doc-1]
+	}
+
+	for len(t.sections) > 0 && t.sections[0].Start < fileStart {
```
Zoekt doesn't seem to be doing any skipping to find the starting rune document section for a file. It's a linear search. Perhaps, we can store additional data (starting rune document section offset per file) in the index file to make this O(1).

keegancsmith replied:
Another format is to group the sections into batches of 1024 (or some other number). This allows you to skip quickly if your search only applies to a few files. This is something I have been exploring for the posting list. But document jumping also makes sense :+1:

keegancsmith reviewed (commented):
This is awesome! I think we can continue this review in upstream zoekt, since it looks good to me! Can you submit that?

> fix unit tests

This will be a requirement before sending it to upstream or landing it locally. Please do this next.

> store corresponding ctags metadata
> expose the metadata in the zoekt api

Save both of these for another PR. However, you can start working on them (just don't bundle them into this PR)

As for next steps, I think we send this upstream. Normally turn around time is good. However, if we haven't gotten in a good place by Tuesday next week lets merge this into our master and get it working.

kzh commented:
Gerrit: https://gerrit-review.googlesource.com/c/zoekt/+/233092

kzh committed:
rough: store and expose ctags metadata in zoekt api

keegancsmith commented:
@kzh I'm reviewing this now. The first two commits look good to me, can you just directly land those into master? Then we should probably do `rough: store and expose ctags metadata in zoekt api` as its own PR. But will review it in this PR as well.

My reasoning for separating out the commits is it will make it easier to merge in upstream.

Additionally can you also update the README and say we now enable ctags.

keegancsmith commented on indexdata.go:
```diff
@@ -26,6 +27,8 @@ import (
 // in memory to search. Most of the memory is taken up by the ngram =>
 // offset index.
 type indexData struct {
+	symbolData
```
nit: don't embed this, make it its own field.

keegancsmith commented on indexdata.go:
```diff
@@ -84,6 +89,34 @@ type indexData struct {
+	symIndex         []uint32
+	symKindContent   []byte
+	symKindIndex     []uint32
+	symMetaData      []byte
```
so all of these []byte slices are mmap backed right?

kzh replied:
Yes, they should be. They're referenced using `readSectionBlob`.

```
func (r *indexData) readSectionBlob(sec simpleSection) ([]byte, error) {
	return r.file.Read(sec.off, sec.sz)
}
```

`file` should be `mmapedIndexFile` on linux.

keegancsmith commented on indexdata.go:
```diff
@@ -84,6 +89,34 @@ type indexData struct {
 }
 
+type symbolData struct {
+	symContent       []byte
```
why are we storing symContent? Isn't that already part of the sections in the file content?

keegancsmith commented on api.go:
```diff
@@ -87,6 +94,8 @@ type LineFragmentMatch struct {
 	// Number bytes that match.
 	MatchLength int
+
+	SymbolInfo *Symbol
```
needs a docstring. Also I am assuming this is nil if it is not from a  symbol search? You should say that.

keegancsmith commented on api.go:
```diff
@@ -77,6 +77,13 @@ type LineMatch struct {
+	Sym        string
+	Kind       string
+	Parent     string
+	ParentKind string
```
I asked this in another PR, but is string the best we can do? Additionally, this needs docstrings with examples.

kzh replied:
I stored this data in the index file as enums, but I will also do this in memory.

keegancsmith commented on build/ctags_test.go:
```diff
@@ -32,7 +32,7 @@ func TestTagsToSections(t *testing.T) {
 		}}
 
-	secs, err := tagsToSections(c, tags)
+	secs, _, err := tagsToSections(c, tags)
```
add a test (or update) the symbols returned.

keegancsmith commented on api.go:
```diff
@@ -77,6 +77,13 @@ type LineMatch struct {
+type Symbol struct {
+	Sym        string
+	Kind       string
+	Parent     string
```
ctags calls this a scope. Why do you call it a Parent? I think we call it a parent in the sourcegraph API, but I am not sure if that is the best name here. WDYT?

kzh replied:
Yeah, I was a little confused with the inconsistency between Zoekt and SG naming. Upstream Zoekt also uses `ParentType` instead of `ScopeKind` which I found strange. I changed it to `ParentKind` in this PR.

s/Parent/Scope and s/Type/Kind are probably the best naming choices here since it is what ctags follows.

keegancsmith commented on go.mod:
```diff
@@ -1,20 +1,30 @@
+	golang.org/x/oauth2 v0.0.0-20190604053449-0f29369cfe45
+	golang.org/x/sync v0.0.0-20190423024810-112230192c58
+	golang.org/x/tools v0.0.0-20190809145639-6d4652c779c4 // indirect
 	gopkg.in/src-d/go-git.v4 v4.8.0
```
whats going on here? I'm surprised this file changed, you don't change any deps.

kzh replied:
Hmm, I think this file changed when I ran `go test ~/Code/Sourcegraph/zoekt/...`. I'll remove this from the PR.

keegancsmith commented on indexbuilder.go:
```diff
@@ -149,9 +149,17 @@ func (s *postingsBuilder) newSearchableString(data []byte, byteSections []Docume
 type IndexBuilder struct {
 	contentStrings  []*searchableString
 	nameStrings     []*searchableString
+	symbolStrings   []*searchableString
```
why is this searchableString? And why are you adding this? I don't think we read this data since we rely on contentStrings.

keegancsmith replied:
Got it. Rather just duplicate that function to make the type clearer.

keegancsmith commented on index_test.go:
```diff
@@ -1715,7 +1715,7 @@ func TestSymbolBoundaryEnd(t *testing.T) {
 }
 
-func TestSymbolAtom(t *testing.T) {
+func TestSymbolSubstring(t *testing.T) {
```
Maybe extend the test for Symbol metadata?

keegancsmith commented on indexbuilder.go:
```diff
@@ -258,14 +269,25 @@ type Document struct {
+}
+
+func (s symbolSlice) Less(i, j int) bool {
+	return s.symbols[i].Start < s.symbols[j].Start
```
I know this applies to the older code as well, but is it possible to have two symbols with the same Start but different End?

keegancsmith commented on indexbuilder.go:
```diff
@@ -195,6 +203,9 @@ func NewIndexBuilder(r *Repository) (*IndexBuilder, error) {
 	b := &IndexBuilder{
 		contentPostings: newPostingsBuilder(),
 		namePostings:    newPostingsBuilder(),
+		fileEndSymbol:   []uint32{0},
```
why does this start with one entry?

keegancsmith replied:
`fileEndSymbol[0] == 0` since its the index of the first symbol in the first document.

keegancsmith commented on indexdata.go:
```diff
@@ -52,6 +55,8 @@ type indexData struct {
 	fileNameIndex   []uint32
 	fileNameNgrams  map[ngram][]uint32
 
+	fileEndSymbol []uint32
```
why do you need fileEndSymbol? Can't we just rely on the same indexes as our document sections in the matching code?

keegancsmith reviewed (commented):
A lot to absorb :) But tried to do a decent review. But I think for the second round of review what is missing for me is some more high level info. So my suggestion (also sent in slack):

I think you should open a new PR for metadata. And include PR description explaining exactly how you are changing what is stored in the index, how we expose this in the UI and/or API, motivation for why you are storing what you are storing, explanation of what you think could be changed/improved, rough estimation on resident memory usage. Additionally some context on how this will relate to upstream is important.

What is also missing for me is seeing the changes you need to make to sourcegraph/sourcegraph for this to be used.

keegancsmith committed:
revert go.mod changes

Change-Id: I8290936305f805032e499361538f00f128153a9d

keegancsmith committed:
Avoid reading symMetaDataIndex

It is fixed size, so we can compute symMetaDataIndex from i. Additionally we
add several comments to the codebase related to symbols.

Change-Id: Ic9c658e79903d345523c32eb3e568d56d3068bc3

keegancsmith committed:
Change symMetaData to a simpleSection

Change-Id: Ie6d1c9381cbf0fc532f8256d9dbed9634cc633a6
